== fails to match types on associated type with two PartialEq<A / B> bounds

rustc 1.12.0-nightly (47b3a9834 2016-07-10)

``` rust
pub struct A;
pub struct B;

pub trait Foo {
    type T: PartialEq<A> + PartialEq<B>;
}

pub fn generic<F: Foo>(t: F::T, a: A, b: B) -> bool {
    t == a && t == b
}
```

``` rust
error: mismatched types [--explain E0308]
  --> a.rs:12:10
   |>
12 |>     t == a && t == b
   |>          ^ expected struct `B`, found struct `A`
note: expected type `B`
note:    found type `A`

error: aborting due to previous error
```

Changing the order of bounds in the associated type declaration changes which `==` causes a spurious "mismatched types" error.

Using UFCS works around the issue:

``` rust
    PartialEq::<A>::eq(&t, &a) && t == b
```

---

The same issue exists with bounds like `type T: PartialEq + PartialEq<B>` to compare `T` with itself as well as another type.


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"bugadani"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

Triage: [no change](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=437781479a408d2a0f90436aa69897df):

```
error[E0308]: mismatched types
 --> src/lib.rs:9:10
  |
9 |     t == a && t == b
  |          ^ expected struct `B`, found struct `A`
  |
  = note: expected type `B`
             found type `A`
```

--- a/src/test/ui/associated-item/associated-item-two-bounds.rs
+++ b/src/test/ui/associated-item/associated-item-two-bounds.rs
@@ -1,0 +1,16 @@
+// This test is a regression test for #34792
+
+// check-pass
+
+pub struct A;
+pub struct B;
+
+pub trait Foo {
+    type T: PartialEq<A> + PartialEq<B>;
+}
+
+pub fn generic<F: Foo>(t: F::T, a: A, b: B) -> bool {
+    t == a && t == b
+}
+
+pub fn main() {}

